### ISSUE TITLE
a way to find how much REAL disk space a file is using.

### ISSUE THREAD

**Issue by hradec** [lizardfs/lizardfs#609]
During my tests, I start to wonder how much REAL disk space a file would take, according to its goal. 

Regarding standard replication, lets say 3 times,  I'm guessing  a 100MB file would take +-100MB from 3 chunkservers. (+- to account for the 64kb chunk size alignment)

This means, if the LizardFS filesystem has 1000M free, after writing the file it goes down to +-700MB, correct? 

Now, what about in a ec(2,1) goal? how much disk space a 100MB will occupy on each of the 3  chunkservers?

I've tried to look for a way to get the REAL disk space a file is using regarding its goal, but couldn't find any.

I think it would be really nice to have mfsfileinfo to return this information for us, specially for testing and defining the best goal for the data regarding to disk space. It also would be REALLY useful to make real predictions of future needs of disk space against available free disk space!

For example, in our studio, we have projects that generate a large number of files, sometimes files with  10GB or even 1TB.  Having a 1TB file  with goal 3 means we need at least 3TB free.

make sense?

**Comment by hradec** [comment]
haa.. one more question regarding goal 3 vs ec(2,1).

With goal 3, we can loose 2 chunkservers and still be able to read the file, since the file is fully present on all of the chunks. 

With goal 2, we can loose 1 chunkservers and still be able to read the file, since the file is fully present on all of the chunks. 

But what about ec(2,1)? Lets say I loose 2 chunks... can I still read the file? 

from what I understood, ec(2,1) allows me to reconstruct the data from the 2 data parts, or AT LEST 1 data and 1 parity.

so, if I loose booth data, in theory I can't reconstruct then from the parity, correct? 

So, in a sense, ec(2,1) would be equivalent to a simple goal 2, but using more disk space with 3 servers instead of 2? 

make sense?

**Comment by eriklindahl** [comment]
The "data parts" in erasure coding isn't the same as replication, but different parts of the data. ec(2,1) roughly means you split chunks two ways, and then add another parity part. This would only use 150MB of storage in your case. If you lose two servers, the file is toast.

The real power of erasure coding comes with slightly higher levels (say, 8+2), where we only waste 25% of the storage but can withstand two drive failures without losing anything.

**Comment by 4Dolio** [comment]
Is it correct to conceptualize erasure as yielding similar tolerances as raid+parity? Though the mechanics are different.

So ec(2,1) is two members and 1 parity? Like raid5 on 2 disks? would be silly to actually use?

Is ec(3,1) similar to 3 member raid5?

And ec(3,2) similar to 3 member raid6? Again silly.

Not untill you have more than 3 members does it begin to pay off?

So ec(8,2) is like 8 member raid6(double parity)

What are the limits of ec parity value?

Is ec(8,4) like 8 member raid8(quad parity if such a thing existed? 50% waste, tolerates 4 lost members?

Then we get into fun stuff like:
ec(16,4) 25% waste, tolerance 4
ec(16,2) 12% waste, tolerance 2
Etc...

**Comment by psarna** [comment]
Yes, it's conceptually similar to raid. With ec(X, Y) you divide each 64MB
chunk into 64/X MB data parts and compute Y parity parts. Any X parts are
enough to read data, so, in other words, you can read even if you lose up
to Y parts. Maximum value for ec goal is ec(32,32). Anything between that
and ec(2,1) is available.

On Saturday, October 7, 2017, 4Dolio <notifications@github.com> wrote:
> Is it correct to conceptualize erasure as yielding similar tolerances as
raid+parity? Though the mechanics are different.
>
> So ec(2,1) is two members and 1 parity? Like raid5 on 2 disks? would be
silly to actually use?
>
> Is ec(3,1) similar to 3 member raid5?
>
> And ec(3,2) similar to 3 member raid6? Again silly.
>
> Not untill you have more than 3 members does it begin to pay off?
>
> So ec(8,2) is like 8 member raid6(double parity)
>
> What are the limits of ec parity value?
>
> Is ec(8,4) like 8 member raid8(quad parity if such a thing existed? 50%
waste, tolerates 4 lost members?
>
> Then we get into fun stuff like:
> ec(16,4) 25% waste, tolerance 4
> ec(16,2) 12% waste, tolerance 2
> Etc...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci4.googleusercontent.com/proxy/lwWNXdOzS9y5aZxakv9EbVlRWww2b2mirF13Bb22aBWTz1AhMenQlbkJoBbdeyEXuDvyVjKmAwuRs7d9nRDENR9CkNvrGso1DpnOVnZjNt4Tfm4pBqaKuaSmmkrjtwbY-eBZ6dsczR1jajbhHambVzUswdwYQg=s0-d-e1-ft#https://github.com/notifications/beacon/AJ8yF6jGraGK-GYvOBxc4FWtNpcPQw5Oks5sp86sgaJpZM4PvgY1.gif>

**Comment by 4Dolio** [comment]
Thanks for the confirmation. And the upper bound of 32, good to know.  I do not yet use or want to use ec, but it is a very interesting feature.  Someday i will play around with it, i do have enough CS members that it might be worthwhile to leverage it.

Since i only use goals of 2 or 3 though there isn't as much advantage for me yet. And my other use cases of having goal=allCS such that each can be whole authoritative doesn't seem worthwhile... Or would is? How would goal=4 compare to ec(4,4)?

**Comment by eriklindahl** [comment]
You should add the data and parity parts, so ec(8,2) would be equivalent to 10-disk raid6.

Both goal=4 and ec(4,4) means your data will survive any 4 failed disks. goal=4 will use 400% of the storage, while ec(4,4) will use 200%, although the access can sometimes be a bit slower.

**Comment by eriklindahl** [comment]
... and the other drawback of ec(4,4) is that you need at least 8 physical servers.

**Comment by 4Dolio** [comment]
So then my raid analogies are not correct in regards to the number of fragments or members.  Thanks, and i should have known.

So is ec literally what psarna described it as (x,y) with x data plus y parity? Or is it more literally (x,y) where y fault encoding method is used to encode x+y fragments? Practically speaking the end result is the same, but i am trying to better grasp the raw mechanics better.  Is is X fragments with Y parity or X+Y fragments encoded for Y redundency?

**Comment by guestisp** [comment]
X data plus Y parity

EC(4,2) is similar to a 6 disks raid6 where your data is split in 4 data parts plus 2 additional parities

You can loose up to 2 disks. If you loose the third, you'll loose data

**Comment by 4Dolio** [comment]
> account for the 64kb chunk size alignment
Chunks are up to 64MB not 64KB i believe.

**Comment by njhurst** [comment]
@eriklindahl actually, you just need at least 8 separate chunkservers.  I'm running 5,3 on just two machines each with many many drives, but with a separate chunkserver per drive.  It works great.  Of course if one of those machines catches fire I'm out of luck...  Also, goal=4 means you have 4 copies, you can only lose 3 drives before you lose data.  

@4Dolio  I don't believe there is any requirement for a chunk to be exactly 64MB (so there's no alignment penalty AFAIK).

**Comment by 4Dolio** [comment]
Shrugs. I was just pointing out the potential typo? Chunks are 'up to' 64MB, they can be smaller of course, but not larger.  They are not 64KB was my point, maybe ec uses different sizes?

**Comment by psarna** [comment]
Just a side note - actually, a 64kB alignment exists in LizardFS as well. Every regular chunk consists of "up to 1024 blocks, each one 64kB in size". For ec(X, Y) chunks, it's "up to 1024/X blocks, each one 64kB in size".
